File: problems/207/solution2.py
from collections import defaultdict


class Solution:
    def canFinish(self, numCourses: int, prerequisites: list[list[int]]) -> bool:
        graph = defaultdict(list)

        for course, prereq in prerequisites:
            graph[prereq].append(course)

        UNVISITED, VISITING, DONE = 0, 1, 2
        state = [UNVISITED] * numCourses

        def dfs(course: int) -> bool:
            if state[course] == VISITING:
                return False
            if state[course] == DONE:
                return True

            state[course] = VISITING
            for prereq in graph[course]:
                if not dfs(prereq):
                    return False
            state[course] = DONE
            return True

        for course in range(numCourses):
            if state[course] == UNVISITED and not dfs(course):
                return False
        return True

File: java/leetcode/src/main/java/com/civoduts/solution/_207/Solution2.java
package com.civoduts.solution._207;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

public class Solution2 {
  public boolean canFinish(int numCourses, int[][] prerequisites) {
    List<List<Integer>> adj = new ArrayList<>(numCourses);
    for (int i = 0; i < numCourses; i++) {
      adj.add(new ArrayList<>());
    }

    int[] indegrees = new int[numCourses];
    for (int[] pair : prerequisites) {
      int course = pair[0];
      int prereq = pair[1];

      adj.get(prereq).add(course);
      indegrees[course] += 1;
    }

    Queue<Integer> queue = new ArrayDeque<>();
    for (int i = 0; i < indegrees.length; i++) {
      if (indegrees[i] == 0)
        queue.offer(i);
    }

    int finished = 0;
    while (!queue.isEmpty()) {
      int curr = queue.poll();
      finished++;

      for (Integer next : adj.get(curr)) {
        if (--indegrees[next] == 0)
          queue.offer(next);
      }
    }

    return finished == numCourses;
  }
}